cpp/src/spatial/hausdorff.cu

```diff
@@ -0,0 +1,189 @@
+    }
+}
+
+struct hausdorff_functor
+{
+    template<typename T, typename... Args>
+    std::enable_if_t<not std::is_floating_point<T>::value, std::unique_ptr<cudf::column>>
+    operator()(Args&&...)
+    {
+        CUSPATIAL_FAIL("Non-floating point operation is not supported");
+    }
+
+    template<typename T>
+    std::enable_if_t<std::is_floating_point<T>::value, std::unique_ptr<cudf::column>>
+    operator()(cudf::column_view const& xs,
+               cudf::column_view const& ys,
+               cudf::column_view const& points_per_space,
+               rmm::mr::device_memory_resource *mr,
+               cudaStream_t stream)
+    {
+        auto tid = cudf::experimental::type_to_id<T>();
+        auto result = cudf::make_fixed_width_column(cudf::data_type{ tid },
+                                                    points_per_space.size() * points_per_space.size(),
+                                                    cudf::mask_state::UNALLOCATED,
+                                                    stream,
+                                                    mr);
+
+        if (result->size() == 0)
+        {
+            return result;
+        }
+
+        auto space_offsets = rmm::device_vector<cudf::size_type>(points_per_space.size());
+
+        thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream),
+                               points_per_space.begin<cudf::size_type>(),
+                               points_per_space.end<cudf::size_type>(),
+                               space_offsets.begin());
+
+        // utilize one block per result (pair of spaces).
+        int max_thread_count = 65535;
+        int block_x = min(result->size(), max_thread_count);
+        int block_y = ceil(result->size() / (float) max_thread_count);
+
+        const uint32_t NUM_THREADS = 1024;
+        dim3 grid(block_x, block_y);
+        dim3 block(NUM_THREADS);
+
+        auto kernel = kernel_hausdorff<T>;
+
+        kernel<<<grid, block, 0, stream>>>(
+            points_per_space.size(),
+            xs.data<T>(),
+            ys.data<T>(),
+            space_offsets.data().get(),
+            result->mutable_view().data<T>()
+        );
+
+        return result;
+    }
+};
+
+} // namespace anonymous
+
+namespace cuspatial {
+
+std::unique_ptr<cudf::column>
+directed_hausdorff_distance(cudf::column_view const& xs,
+                            cudf::column_view const& ys,
+                            cudf::column_view const& points_per_space,
+                            rmm::mr::device_memory_resource *mr)
+{
+    CUSPATIAL_EXPECTS(xs.type() == ys.type(), "inputs `xs` and `ys` must have same type.");
+    CUSPATIAL_EXPECTS(xs.size() == ys.size(), "inputs `xs` and `ys` must have same length.");
+
+    CUSPATIAL_EXPECTS(not xs.has_nulls() and
+                      not ys.has_nulls() and
+                      not points_per_space.has_nulls(),
+                      "inputs must not have nulls.");
+
```

Comment:
```suggestion
CUSPATIAL_EXPECTS(points_per_space.size() < 65536, "points_per_space must be less than 65536");
```
Or similar, otherwise `points_per_space.size()` squared will overflow `size_type`

Reply:
Oh, actually this should be `< 46341` since `size_type` is signed too...
Would `constexpr auto too_big = size_type(ceil(sqrt(std::numeric_limits<size_type>::max())));` be too much of an atrocity?

Reply:
It'd be `floor`, thought right?

Reply:
I'll take a closer look to see what we can do.

Reply:
`floor` would give you the largest safe value, so you could use it with `<=`

Reply:
Please avoid hard-coding an integer literal...  What is the source of 65536 here?

Reply:
I wasn't sure, which is why I left it as a literal (hoping this conversation would happen - but forgetting to add a comment myself). It's related to the number of blocks we can have in a grid, I think? Not sure what to name it.

Reply:
> Please avoid hard-coding an integer literal... What is the source of 65536 here?

65536 * 65536 is the smallest square that would overflow a 32-bit unsigned int. But then I remembered that `size_type` is signed. That's why I suggested `size_type(ceil(sqrt(std::numeric_limits<size_type>::max())))` instead (which, thanks to `constexpr`, can be evaluated at compile time)

Reply:
I'm leaving this out for the time being, because I do not have a / do not know how to verify this. Instead, I'm documenting what will happen in the event too many spaces are passed to the algorithm.

Reply:
Ended up adding a constant, commenting it, and using it as a guard: `points_per_space.size() > MAX_NUM_SPACES`.